Upgrade/v1.79

KinoxKlark opened the pull request:
I started to upgrade imgui from v1.65 to v1.79

* Removed implemented functions that are not part of ImGui anymore (DisplayVisibleMin, DisplayVisibleMax)
* Raname ConfigResizeWindowsFromEdges to ConfigWindowsResizeFromEdges
* Marked obsolete functions as such
* Cross checked both API and added references to unimplemented functions from the new version
    * I also reordered function and added comments such that it resembles imgui.h as possible in order to facilitate upgrades from here
* Updated all enums

It is not finished yet, there is a lot of new unimplemented functions, but it should pass tests and be usable with the new ImGui code. I quickly tested the imgui.show_test_window(), and new features seem to work fine. I am willing to put some time into mapping functions to complete the upgrade. Feedback would be great :)

swistakm commented:
OH MY GOD. That's huge! And great!

How did you manage to do all of that? Last time I was attempting the update I have realized how many flags there are and how often they change I had to create some [custom tooling to automate that stuff](https://github.com/swistakm/pyimgui/tree/imgui-upstream-v1.77-branch/update-tools).

swistakm commented:
I'm totally loaded with work at the moment but I quit my current job next week so finally will have more time to do all the reviews, merges and releases. Anyway, don't stress about missing functions. We never had 100% feature parity (imgui has HUGE API). Just tell me when you're ready and we merge it. Later we can do incremental updates.

KinoxKlark commented:
> OH MY GOD. That's huge! And great!
> 
> How did you manage to do all of that? Last time I was attempting the update I have realized how many flags there are and how often they change I had to create some [custom tooling to automate that stuff](https://github.com/swistakm/pyimgui/tree/imgui-upstream-v1.77-branch/update-tools).

I saw your scripts but was not really sure of how to use them properly. I initially wanted to code my own and even started to list all features that such a script should have to compare versions and propose updates, but I figured out that I needed to go manually through the entire process once before being able to automatize it. Maybe I should regroup my notes and take the time to understand your scripts to plan the development of a generic tool that we could use after each update of ImGui. A semi-automated process that checks changes between ImGui versions and proposes the modifications to do would be of great help for future upgrades.

swistakm commented:
@KinoxKlark BTW sorry for cancelling your builds on travis. I'm working on a new Appveyor pipeline for macOS builds because the old one on travis became unmanageable. Appveyor isn't perfect but simply works better and faster. Maybe will migrate the manylinux1 build to Appveyor too. When I get it done I'll ask you to rebase on master. OK?

KinoxKlark commented:
@swistakm Ok no problems ;)

swistakm commented:
Another note: I've noticed that after I've merged the #167 your builds may fail after the rebase. I think that this patch should be sufficient:
```diff
diff --git a/ansifeed-cpp/AnsiTextColored.cpp b/ansifeed-cpp/AnsiTextColored.cpp
index f34486d..4022a75 100644
--- a/ansifeed-cpp/AnsiTextColored.cpp
+++ b/ansifeed-cpp/AnsiTextColored.cpp
@@ -116,8 +116,8 @@ namespace ImGui
                                                          // so this is merely to handle direct calls.
 
         // Align to be pixel perfect
-        pos.x = (float)(int)pos.x + font->DisplayOffset.x;
-        pos.y = (float)(int)pos.y + font->DisplayOffset.y;
+        pos.x = (float)(int)pos.x;
+        pos.y = (float)(int)pos.y;
         float x = pos.x;
         float y = pos.y;
         if (y > clip_rect.w)
@@ -447,7 +447,7 @@ namespace ImGui
         if (text_end == NULL)
             text_end = text + strlen(text);  // FIXME-OPT
 
-        const ImVec2 text_pos(window->DC.CursorPos.x, window->DC.CursorPos.y + window->DC.CurrentLineTextBaseOffset);
+        const ImVec2 text_pos(window->DC.CursorPos.x, window->DC.CursorPos.y + window->DC.CurrLineTextBaseOffset);
         const float wrap_pos_x = window->DC.TextWrapPos;
         const bool wrap_enabled = wrap_pos_x >= 0.0f;
         if (text_end - text > 2000 && !wrap_enabled) {
```
You should be able to simply `git apply` it.

Johndeep commented:
Hopefully this draft pull request will make it in this year still. At least that's how this pull request looks like. I heavily make use of pyimgui in Python 3.9 and it would be awesome to be finally able to use the new docking layout system and tabbing and more. So many thanks for this already :+1:

KinoxKlark commented:
> Hopefully this draft pull request will make it in this year still. At least that's how this pull request looks like. I heavily make use of pyimgui in Python 3.9 in Debian and it would be awesome to be finally able to use the new docking layout system and tabbing and more. So many thanks for this already 👍

Thanks, that is motivating. Unfortunately, I can't promise anything for this year. Exams are coming, and I will soon have less time. I try to do as much as I can for now.

Concerning docking and viewport: As I understand it, it is still a working branch in the original imgui project. It is safe to use, but I am wondering if we should add it to master. Maybe a good idea would be to also create a docking branch on pyimgui that we would keep up to date. I don't know how it would work with pip.

KinoxKlark commented:
I have reviewed the list of releases from 1.65 to 1.79 in order to apply all changes, especially breaking ones. Easily mapped functions have been mapped to.

Here a more detailed list of changes for this pull request :

* Checkout Imgui v1.79 branch
* Update of the imported API in cimgui.pxd
* Update of all enums and flags and addition to doc (documentation not finished, just imported)
* Mapping of all enums and flags
* Mapping of functions:
    * ``SetNextWindowSize()``
    * ``SetScrollHereX()``
    * ``SetScrollHereY()``
    * ``SliderAngle()``
    * ``GetDragDropPayload()``
    * ``ShowAboutWindow()``
    * ``BeginTabBar()`` <- Tab features fully usable :D
    * ``EndTabBar()``
    * ``TabItemButton()``
    * ``SetTabItemClosed()``
    * ``ImDrawData::DisplayPos``
    * ``ImDrawData::DisplaySize``
    * ``ImDrawData::FramebufferScale``
    * ``IsItemEdited()``
    * ``IsItemActivated()``
    * ``IsItemDeactivated()``
    * ``IsItemDeactivatedAfterEdit()``
    * ``IsItemToggledOpen()``
    * ``ImGuiStyle::LogSliderDeadzone``
    * ``ImGuiStyle::TabRounding``
    * ``ImGuiStyle::TabBorderSize``
    * ``ImGuiStyle::TabMinWidthForCloseButton``
    * ``ImGuiStyle::SelectableTextAlign``
    * ``ImGuiStyle::AntiAliasedLinesUseTex``
    * ``ImGuiStyle::CircleSegmentMaxError``
    * ``GetBackgroundDrawList()``
    * ``InputTextWithHint()``
    * ``SetNextItemWidth()``
    * ``SetNextItemOpen()``
    * ``GetTreeNodeToLabelSpacing()``
    * ``ImGuiStyle::ColorButtonPosition``
    * ``ImGuiStyle::WindowMenuButtonPosition``
    * ``ImDrawList::AddNgon()``
    * ``ImDrawList::AddNgonFilled()``
    * ``AddInputCharacterUTF16()``
    * ``OpenPopupOnItemClick()``
    * ``BeginPopupContextVoid()``
    * ``IsPopupOpen()``
    * ``GetGlyphRangesThai()``
    * ``GetGlyphRangesVietnamese()``
    * ``ImGuiIO::MetricsRenderIndices``
    * ``ImGuiIO::MetricsRenderWindows``
    * ``ImGuiIO::MetricsActiveAllocations``
* Functions that have changed of prototype:
    * ``SliderXXX()`` now use ``SliderFlags`` and no ``power`` parameters. ``power`` is kept as it is still available marked as OBSOLETED in imgui API.
    * ``DragXXX()`` now use ``SliderFlags`` and no ``power`` parameters. ``power`` is kept as it is still available marked as OBSOLETED in imgui API.
    * ``DragXXX()``
    * ``AddInputCharacter()`` change signature from ``unsigned short c`` to ``unsigned int c``
    * ``AddRectXXX()`` change flags default value to use enum instead of direct number
    * ``AddCicle()`` argument ``num_segment`` now defautls to 0 meaning auto-tesselation
    * ``AddCicleFilled()`` argument ``num_segment`` now defautls to 0 meaning auto-tesselation
    * ``IsKeyPressed()`` first argument should not default anymore
    * ``IsItemClicked()`` Arguments type change from int to ImGuiMouseButton
    * ``BeginPopupContextWindow()``
    * ``InvisibleButton()`` Now take a ImGuiBttonFlags in input
    
As far as I know, it is now fully upgraded to 1.79. Missing mappings were already missing in the old version. New features were added. The list of features to map in ``cimgui.pxd`` should be up to date with ``imgui.h`` such that it is now easy to see which features miss. I documented new mapping as I implemented them, but it is not complete nor perfect.

Tests should be done to check that there are no mistakes. I quickly tested each new feature but not all possible cases. Documentation should be updated too. I saw mixed references to pyimgui and imgui API in docstrings, but I did not fix it each time. Documentation does not seams to be normalized. I think a massive pass on the documentation should be done. I don't know if all of this should be done in this branch or if we start by merging it to master and plan afterward?

**Concerning the next upgrades:** v1.79 is the most recent release of ImGui but not up to date with the master branch. It would be nice to have a clear idea of which versions to map in which branch of this project? Should we have an imgui/v1.79 branch with a clean mapping and keep the master branch up mapped to ImGui's master? Should we have a docking branch? I think a mirroring of ImGui's most important branching could be great. It would then be possible to map features as they are appended and changed in the ImGui repo. This way, we could be fast to upgrade when ImGui updates. What do you think?

swistakm commented:
Wow
<img width="134" alt="Zrzut ekranu 2020-12-16 o 11 23 53" src="https://user-images.githubusercontent.com/1258054/102336574-a2426480-3f91-11eb-9940-784582b412bd.png">
That is awesome!

Regarding the branching. My initial idea was to simply merge your changes to master and release the changes as 2.0.0. Supporting all past upstream ImGui versions isn't worth the effort in my opinion. 

I don't know what to do about docking branch yet. It is synced with master systematically so we could do the same and release it also on a branch synced with our master. But I'm not sure what to do with PyPI releases. That would take another package name for sure. If we want to do this we would have also to invest in faster build pipelines, because at current situation it takes too much of time to get all the build artifacts. That could also create some confusion if user installs both imgui and e.g. imgui-docking from PyPI.

What we can do for sure is to setup separate branch experimentally for docking feature and see how it goes. Maybe users interested in that particular feature can handle installing from sources until docking becomes official part of upstream release. Still, I'm afraid scaling up to more extra features on branches won't be sustainable.

I'm also not sure what to do with upstream master.

Johndeep commented:
About the master branch or the stable release branch. In my humble opinion, we should stick to the stable release for now, since the master is per say the alpha, ever changing.

If going with the master, I assume that this means, you have to correlate every little change every now and then (worst case, every day), since the master branch changes quite often, in contrast to the stable one. Furthermore, if you don't want every little change, how can you know, which change to adapt to be sure the package is suddenly not broken somewhere (regressions)?
On the other hand, the stable release shouldn't change until the next stable release, I think. Since the pyimgui version is an everyday version for us to work with regularly, it would be reassuring to have it based on something recent but stable. As @swistakm also mentioned, the resources for recompiling everything regularly is limited.

Thus, I would already be very happy and be quite satisfied to having pyimgui be at the same version as their stable though.

KinoxKlark commented:
Yeah, you make both excellent points.

I changed the config of this pull request from 'draft' to 'ready for review.' The upgrade is not entirely finished because all features are not fully mapped, but it was never the case as I understand it. Maybe we should wait a little more until releasing it as the 2.0.0 to have time to test it and complete it a little more. It would be nice to have a clean version for the release. The reason why I set it ready to review is to "officialize" a little bit more this work as a real work in progress, and maybe you want to merge it to a working branch such that other people start to work on it too?

 As far as I know, the upgrade to 1.79 is done since it compiles fine with the imgui-cpp 1.79 branch, and it contains numbers of the new features (e.g. TabBar). I am currently working on some clipboard stuff, new functions to map that break the existing callback system. But I have the impression that I could endlessly work on implementing and fixing features. Maybe it would be time to separate the work features by features?

Concerning branches, my proposal is the following: Indeed, matching every feature branch of imgui is a bit too much and would be too much work and effort. @Johndeep makes good points concerning imgui's master branch. Limiting our work to official releases is probably the best choice. I was afraid that the effort necessary to upgrade after each release would be too big and take too much time to do. But if it is done rigorously after each update with a standardized process (even with semi-automated scripts), I think that it is feasible (more feasible than being up to date with every little change of the master branch indeed). Concerning the docking branch it seems to be a feature that has great interest. Part of the reason why I started this upgrade is that I, myself, wanted to use docking features and found that for implementing the docking branch a first step was to upgrade to the latest release. Thus a pyimgui/docking experimental branch would be great. Interested people could simply install from source for now. 

What do you think about the next concerning this upgrade? Should we plan features implementation, testing, and documentation redaction? Should this branch be integrating to pyimgui for community work, or should I keep my current fork working on it?

KinoxKlark commented:
It seems that there is a need for mapping some of the ``imgui_internal.h`` functionality like ``PushItemFlag()`` as in #178. Should we provide this? I don't think it has its place in ``pyimgui.core`` but maybe in a ``pyimgui.internal`` module? What do you think?

swistakm commented:
> It seems that there is a need for mapping some of the imgui_internal.h functionality like PushItemFlag() as in #178. Should we provide this? I don't think it has its place in pyimgui.core but maybe in a pyimgui.internal module? What do you think?

👍 from me for both exposing some internals and doing that in separate module. I just wonder if that would be easy to do as a separate pyx module? I remember that I wanted to modularize the library at the very beginning but had some technical troubles with that. I've eventually put everything in core and just prefixed stuff that should be publicly exposed in different submodule (see extras). If you know how to do that properly, you're more than welcome to do that :)

KinoxKlark commented:
> 👍 from me for both exposing some internals and doing that in separate module. I just wonder if that would be easy to do as a separate pyx module? I remember that I wanted to modularize the library at the very beginning but had some technical troubles with that. I've eventually put everything in core and just prefixed stuff that should be publicly exposed in different submodule (see extras). If you know how to do that properly, you're more than welcome to do that :)

Indeed it wasn't easy, but I think I managed to do it. Maybe you should verify that it works as intended and that I didn't forget anything. Adding the internal submodule wasn't really difficult, just adapting the `setup.py` and finding a way to use common code between `core` and `internal` like the error handling. The main difficulty was that both modules share different imgui context. The solution I found was to propagate the change in context trough submodules using a `UpdateImGuiContext()` calling `create_context()`, `destroy_context()` and `get_current_context()`. I don't think that there is any other global state that every submodule should be aware of, but I may be wrong. If you think that there is a better way to do it, let me know ^^

To test it I mapped `PushItemFlag()` and `PopItemFlag()` from internal and tested the disabled features asked in  #178:
```python
disabled = False

# ...

imgui.begin("Form")
clicked, disabled = imgui.checkbox("Disable", disabled)
if disabled:
    imgui.internal.push_item_flag(imgui.internal.ITEM_DISABLED, True)
    imgui.push_style_var(imgui.STYLE_ALPHA, imgui.get_style().alpha * 0.5)

imgui.text('This is a test')
if imgui.button('Click me!'):
    print('Clicked')

if disabled:
    imgui.internal.pop_item_flag()
    imgui.pop_style_var()
    
imgui.end()
```

TTimo commented:
Getting this error at startup on occasion:

```
  File "imgui/core.pyx", line 2431, in imgui.core.new_frame
imgui.core.ImGuiError: ImGui assertion error ((g.IO.DeltaTime > 0.0f || g.FrameCount == 0) && "Need a positive DeltaTime!") at imgui-cpp/imgui.cpp:6867
```

This was happening in https://github.com/swistakm/pyimgui/pull/162 as well, I haven't tested 1.3.0 but I'm pretty sure it never happened in 1.2

KinoxKlark commented:
> Getting this error at startup on occasion:
> 
> ```
>   File "imgui/core.pyx", line 2431, in imgui.core.new_frame
> imgui.core.ImGuiError: ImGui assertion error ((g.IO.DeltaTime > 0.0f || g.FrameCount == 0) && "Need a positive DeltaTime!") at imgui-cpp/imgui.cpp:6867
> ```
> 
> This was happening in #162 as well, I haven't tested 1.3.0 but I'm pretty sure it never happened in 1.2

Thanks, I'll look at it! 

Unfortunately, I won't have a lot of time in the next few days since my exam session started at university. I'll come back to this dev mid January :)

zhihaoy commented:
`IsItemDeactivatedAfterEdit` is great, otherwise `input_text_multiline` is very hard to use.

KinoxKlark commented:
> `IsItemDeactivatedAfterEdit` is great, otherwise `input_text_multiline` is very hard to use.

Thanks for the feedback! Concerning the difficulty using `input_text_multiline`, can you be a little more precise about the problem you encounter? Is it the `buffer_length` input parameter that does not feel natural? If so, maybe we should set it to a default value, but I am not sure which one yet. I think it is quite usage dependent.

zhihaoy commented:
> > `IsItemDeactivatedAfterEdit` is great, otherwise `input_text_multiline` is very hard to use.
> 
> Thanks for the feedback! Concerning the difficulty using `input_text_multiline`, can you be a little more precise about the problem you encounter?

You usually need some way to "commit" an input line. I can add a button to do so, but I prefer pressing the Enter key. But you can't achieve this behavior easily with `input_text_multiline`, because the default behavior is to return `True` if "changed." Setting the `INPUT_TEXT_ENTER_RETURNS_TRUE` flag works for `input_text` but has the opposite meaning for `input_text_multiline` -- Enter will open a new line and Ctrl-Enter commit returns True. `input_text_multiline` needs to be used with `IsItemDeactivatedAfterEdit` if you want to let Enter key to commit the change.

zhihaoy commented:
Oh, nvm. I didn't notice that `imgui.INPUT_TEXT_CTRL_ENTER_FOR_NEW_LINE` is a different flag.

KinoxKlark commented:
Indeed `imgui.INPUT_TEXT_CTRL_ENTER_FOR_NEW_LINE` should inverse the default comportment and permits unfocus with `Enter` and create a new line with `Ctrl+Enter`, which sound like what you want to do. This is a clear case where more documentation is needed, maybe with example cases such as this one for the different flags.

KinoxKlark commented:
Finally, my exam session is done, and I will have more time to invest in this project. 

Big news, imgui release 1.80 is here, what should we do about that? For now, I continue with 1.79 but should we directly update to 1.80 for the 2.0.0 release? I quickly checked the changelog: the main new thing is the merge of Table in master branch, a feature that has been requested for pyimgui here #196. There are also various renames and minor breaking changes that should be dealt with but nothing to complicate it seems.

Johndeep commented:
In my humble opinion, if the jump to version 1.80 does not add too much work, it would be advisable to go with 1.80, also because it is now the current stable and there should be many bug fixes and of course the table feature.
Unfortunately, there is still no docking feature though. But if we would at least can raise the minimum version, the feature gap between pyimgui and imgui official can be kept to a minimum.

You mentioned that there are "also various renames and minor breaking changes". Does this refer to the pyimgui v1.3 or the jump from 1.79 to 1.80 from imgui official?

KinoxKlark commented:
> You mentioned that there are "also various renames and minor breaking changes". Does this refer to the pyimgui v1.3 or the jump from 1.79 to 1.80 from imgui official?

I am referring to the jump from 1.79 to 1.80 of imgui. As in each release, they have a section of their changelog dedicated to breaking changes from the previous release to the new one. I didn't get through it in detail, but it is usually quick fixes like function renaming and deletion of old obsolete stuff.

From pyimgui 1.3.0 to 2.0.0, they will be breaking changes as a lot of stuff has been moved or renamed during this massive upgrade, without mentioning the drop of support for old python versions. I think the idea, for now, is to get as close as possible to the official imgui and for future release be more carful about breaking changes and how we want to deal with them.

KinoxKlark commented:
> Getting this error at startup on occasion:
> 
> ```
>   File "imgui/core.pyx", line 2431, in imgui.core.new_frame
> imgui.core.ImGuiError: ImGui assertion error ((g.IO.DeltaTime > 0.0f || g.FrameCount == 0) && "Need a positive DeltaTime!") at imgui-cpp/imgui.cpp:6867

@TTimo: I think it should be ok now. Negative io.delta_time has been made illegal in imgui 1.68. To fix it, I had to change integration codes, checking for a negative delta time and fixing delta_time to a small positive value when necessary (I think the problem generally arises at startup when delta_time == 0). Depending on the backend, I had to add a `process_inputs()` function that has to be called every frame. It was already like that for `gfwl` and `pysdl2`; I added it to `pygame`, `pyglet`, and `coco2d`. Example code has been updated. Let me know if there are any problems.

TTimo commented:
@KinoxKlark sounds good! I've never seen the assert fire with DeltaTime being negative, only it being 0

tpwrules commented:
Could you please add the NavInputs array so I can implement gamepad navigation?

KinoxKlark commented:
> Could you please add the NavInputs array so I can implement gamepad navigation?

@tpwrules, `IO.NavInputs` should now be mapped as `io.nav_inputs`. `ConfigFlags` and `BackendFlags` were already mapped as `config_flags` and `backend_falgs`. Please, let me know if there are any issues. I didn't fully test it. I didn't either modify the backends integration code, so if you feel like sharing your implementation of gamepad navigation, it would be great!

tpwrules commented:
@KinoxKlark You mis-spelled the `NAV_INPUT_TEWAK_FAST` constant :) Other than that, it works perfect. Thank you!

My joystick code is nothing special. Just a loop each frame that reads the relevant buttons using pygame's joystick API and updates `nav_inputs` appropriately.

KinoxKlark commented:
I found a way to implements callbacks as requested in #200.

Python callable and user data are bundled together in a `_callback_user_info` object, which is used as `user_data` in imgui API. A generic callback C function is used as the function pointer, and this function is in charge of calling the desired python callable and passing it correct user data. I don't know if this is the best way to do it, but it seems to work fine for InputTextCallback and SizeCallback.

swistakm commented:
> Finally, my exam session is done, and I will have more time to invest in this project.

Great to hear that. I hope all the exams went well.

> Big news, imgui release 1.80 is here, what should we do about that? For now, I continue with 1.79 but should we directly update to 1.80 for the 2.0.0 release?

I agree with @Johndeep. I think that if the amount of changes isn't that big and you are able to cover them then it is really worth update straight to 1.80. Feel free to make that stretch if you have time for that :) Especially if we can fit all those minor breaking changes directly to the 2.0.0 release.

> Python callable and user data are bundled together in a _callback_user_info object, which is used as user_data in imgui API. A generic callback C function is used as the function pointer, and this function is in charge of calling the desired python callable and passing it correct user data. I don't know if this is the best way to do it, but it seems to work fine for InputTextCallback and SizeCallback.

Yeah, those unfortunate callbacks... I will have to take a closer look at the code because despite it looks easy it can easily blow up. Because due to how python references on C-level work and how ImGui runs those callbacks it is extremely easy to cause segmentation fault. 

We've already discussed how to properly handle callbacks but never implemented the code for them. You can take a look on #93 where we identified possible problems that may occur and the ways to mitigate them. In the meantime I will try to checkout your code and run some experiments on it.

KinoxKlark commented:
> Great to hear that. I hope all the exams went well.

Thanks, indeed it went well.

> I agree with @Johndeep. I think that if the amount of changes isn't that big and you are able to cover them then it is really worth update straight to 1.80. Feel free to make that stretch if you have time for that :) Especially if we can fit all those minor breaking changes directly to the 2.0.0 release.

I agree too, so I will stretch this upgrade to 1.80 directly.

> Yeah, those unfortunate callbacks... I will have to take a closer look at the code because despite it looks easy it can easily blow up. Because due to how python references on C-level work and how ImGui runs those callbacks it is extremely easy to cause segmentation fault.
> 
> We've already discussed how to properly handle callbacks but never implemented the code for them. You can take a look on #93 where we identified possible problems that may occur and the ways to mitigate them. In the meantime I will try to checkout your code and run some experiments on it.

Thanks for the sharing. With a quick look, I see that I have forgotten to consider lambdas. I will take a more in-depth look at this discussion and run more experiments. Indeed it quickly became more difficult than expected; having some feedback and crosschecking on my implementation would be great!

KinoxKlark commented:
I added various tests for these callbacks in `tests/test_crazy_callbacks.py`. I tried to include as many cases as I could think of; feel free to complete it. I probably forgot many cases.

For now, I tested:
* `input_text` callbacks are correctly called none, one or multiple time depending on events
* If multiple call to `set_next_window_size_constraints()` before `begin()` only the most recent one is kept and memory should not leak (to triple check)
* Callbacks set with `set_next_window_size_constraints()` are only valid for the next `begin()`
* Function as a callback
* Class as a callback
* Class static method as a callback
* Class method as a callback
* Lamba directly in argument as a callback
* Lambda generated by external function as a callback
* Closure as a callback
* Decorated function as a callback

I hope it will help.

KinoxKlark commented:
Update from 1.79 to 1.80 has been done!

Sorry for the branch delete/restore; I wanted to rename the branch to upgrade/v1.80 without success ^^

* I have updated the whole API to be in sync with imgui 1.80
* New flags have been implemented
* Various small changes (renaming, etc)
* I mapped the new Table API and ran some tests. It seems to work fine, but more tests should be done. I didn't do the documentation yet.

Let me know if you see any problems. I will continue testing and making sure everything is okay.

Concerning the Table API, for sorting one has to call `TableGetSortSpecs()` which returns an `ImGuiTableSortSpecs`. This object contains a pointer to `ImGuiTableColumnSortSpecs` that shall be used as an array to the selected columns to sort. I implemented it such that this given field can be used as a python array or an iterable. The syntax is as close as possible as the original Imgui syntax, i.e. `table_sort_specs.specs[idx]` and `table_sort_specs.specs_count`, but I think it is a bit verbose and directly using `table_sort_specs` as an array would be simpler. Like: `table_sort_specs[idx]` and `table_sort_specs.count`. What do you think?

learn-more commented on imgui/integrations/pyglet.py:
```diff
@@ -84,7 +88,7 @@ def _on_mods_change(self, mods, key_pressed = 0):
         self.io.key_shift = mods & key.MOD_SHIFT or \
                             key_pressed in (key.LSHIFT, key.RSHIFT)
-        print(self.io.key_ctrl, self.io.key_super, self.io.key_alt, self.io.key_shift)
+        #print(self.io.key_ctrl, self.io.key_super, self.io.key_alt, self.io.key_shift)
```
This is removed in master

KinoxKlark replied:
Thank you!

learn-more commented on imgui/enums.pxd:
```diff
@@ -217,91 +272,267 @@ cdef extern from "imgui.h":
-        ImGuiMouseCursor_ResizeEW               # When hovering over a column
-        ImGuiMouseCursor_ResizeNESW             # Unused
-        ImGuiMouseCursor_ResizeNWSE             # When hovering over the bottom-right corner of a window
-        ImGuiMouseCursor_Count_
+        ImGuiMouseCursor_None 
+        ImGuiMouseCursor_Arrow 
+        ImGuiMouseCursor_TextInput,         # When hovering over InputText, etc.
+        ImGuiMouseCursor_ResizeAll,         # (Unused by Dear ImGui functions)
+        ImGuiMouseCursor_ResizeNS,          # When hovering over an horizontal border
+        ImGuiMouseCursor_ResizeEW,          # When hovering over a vertical border or a column
+        ImGuiMouseCursor_ResizeNESW,        # When hovering over the bottom-left corner of a window
+        ImGuiMouseCursor_ResizeNWSE,        # When hovering over the bottom-right corner of a window
+        ImGuiMouseCursor_Hand,              # (Unused by Dear ImGui functions. Use for e.g. hyperlinks)
+        ImGuiMouseCursor_NotAllowed,        # When hovering something with disallowed interaction. Usually a crossed circle.
+        ImGuiMouseCursor_COUNT
+
```
This might conflict with #209 if that is merged before.

KinoxKlark replied:
Thanks! Stop me if I'm wrong but I think it should be ok, the only conflicting names are `MOUSE_CURSOR_NONE = enums.ImGuiMouseCursor_None` and `MOUSE_CURSOR_HAND = enums.ImGuiMouseCursor_Hand` and both versions are identical. Thus the conflict resolution should be trivial no ?

learn-more replied:
Indeed.

KinoxKlark commented:
A new release of Imgui is here. I stretched the upgrade to v1.81. This implies:
* `ListBoxHeader()` is now obsolete and `BeginListBox()` should be used. Thus I added `begin_list_box()` and keept `list_box_header()` as a redirection and documented that this function is now obsolete.
* `ListBoxFooter()` is new obsolete and `EndListBox()` should be used. Thus I added `end_list_box()` and keept `list_box_footer()` as a redirection and documented that this function is now obsolete.
* Added new flags `ImGuiViewportFlags_` mapped as `VIEWPORT_FLAGS_NONE`, `VIEWPORT_FLAGS_IS_PLATFORM_WINDOW`, `VIEWPORT_FLAGS_IS_PLATFORM_MONITOR`, and `VIEWPORT_FLAGS_OWNED_BY_APP`
* Mapped new struct `ImGuiViewport`
* Mapped new function `ImGuiViewport* GetMainViewport()` as `get_main_viewport()`

This release implies some changes concerning `imgui_freetype` but I am not sure if it concerns us or not?

learn-more commented:
> A new release of Imgui is here. I stretched the upgrade to v1.81. This implies:
> 
> * `ListBoxHeader()` is now obsolete and `BeginListBox()` should be used. Thus I added `begin_list_box()` and keept `list_box_header()` as a redirection and documented that this function is now obsolete.
> * `ListBoxFooter()` is new obsolete and `EndListBox()` should be used. Thus I added `end_list_box()` and keept `list_box_footer()` as a redirection and documented that this function is now obsolete.
> * Added new flags `ImGuiViewportFlags_` mapped as `VIEWPORT_FLAGS_NONE`, `VIEWPORT_FLAGS_IS_PLATFORM_WINDOW`, `VIEWPORT_FLAGS_IS_PLATFORM_MONITOR`, and `VIEWPORT_FLAGS_OWNED_BY_APP`
> * Mapped new struct `ImGuiViewport`
> * Mapped new function `ImGuiViewport* GetMainViewport()` as `get_main_viewport()`
> 
> This release implies some changes concerning `imgui_freetype` but I am not sure if it concerns us or not?

There is a goal to deprecate the concept of a main viewport in the docking branch:
https://github.com/ocornut/imgui/issues/1542#issuecomment-571121789
So it might be better to not even add this to the api?

KinoxKlark commented:
> There is a goal to deprecate the concept of a main viewport in the docking branch:
> [ocornut/imgui#1542 (comment)](https://github.com/ocornut/imgui/issues/1542#issuecomment-571121789)
> So it might be better to not even add this to the api?

I am not sure to agree, as I understand it `GetMainViewport()` has been added to the `master` branch in order to homogenize `master` and `docking` branches. Indeed the plan seems to transition to a "No main viewport" in the future, but I don't think this invalidates the `Viewport` struct which I suppose will just be used differently. Furthermore, we have no idea in how long this change will come and one of the ideas after completing this upgrade was to map in another branch the much desired `docking` functionalities. Shouldn't we ensure that both API works similarly in that regards if this has been the choice made in official Dear Imgui?

learn-more commented:
> > There is a goal to deprecate the concept of a main viewport in the docking branch:
> > [ocornut/imgui#1542 (comment)](https://github.com/ocornut/imgui/issues/1542#issuecomment-571121789)
> > So it might be better to not even add this to the api?
> 
> I am not sure to agree, as I understand it `GetMainViewport()` has been added to the `master` branch in order to homogenize `master` and `docking` branches. Indeed the plan seems to transition to a "No main viewport" in the future, but I don't think this invalidates the `Viewport` struct which I suppose will just be used differently. Furthermore, we have no idea in how long this change will come and one of the ideas after completing this upgrade was to map in another branch the much desired `docking` functionalities. Shouldn't we ensure that both API works similarly in that regards if this has been the choice made in official Dear Imgui?

You are right here :)
I asked this in the 'Dear ImGui' discord, and this was the reply:

> That question was actually answered by a commit done last week, since GetMainViewport() was introduced to master branch.
> In "no main viewport" mode the bounds returned by GetMainViewport() will likely refer to a monitor
> So as per linked pyimgui's thread, the linked comment is sort of invalidated by this new direction
> Hence adding this function to master
> KinoxKlark's answer is right there
> The definition of ImGuiViewport evolved:
```
enum ImGuiViewportFlags_
{
    ImGuiViewportFlags_IsPlatformWindow         = 1 << 0,   // Represent a Platform Window
    ImGuiViewportFlags_IsPlatformMonitor        = 1 << 1,   // Represent a Platform Monitor (unused yet)
    [....]
};

// - Currently represents the Platform Window created by the application which is hosting our Dear ImGui windows.
// - In 'docking' branch with multi-viewport enabled, we extend this concept to have multiple active viewports.
// - In the future we will extend this concept further to also represent Platform Monitor and support a "no main platform window" operation mode.
// - About Main Area vs Work Area:
//   - Main Area = entire viewport.
//   - Work Area = entire viewport minus sections used by main menu bars (for platform windows), or by task bar (for platform monitor).
//   - Windows are generally trying to stay within the Work Area of their host viewport.
struct ImGuiViewport
{
```

swistakm commented:
Hey guys (and especially @KinoxKlark) I'm really sorry for stalling with imgui but I still under a lot of load from other commitments (book, job, side-gig) and couldn't find even time to respond to latest issues/questions/PRs. I definitely didn't give up on the project but can't give it any attention this month. Maybe April will be more lucky to me.

Anyway, I know that this will be reoccurring problem and I will be ever-busy always. If anyone want's to step in and take at least a part of the responsibility for the project stewardship I will be more than happy to welcome official maintainers with repo write access and proverbial "keys to the castle". The best if we could have trio passionate contributors that would be able to organize themselves, set their direction, decide on things and further drive the project.

KinoxKlark commented:
Hey @swistakm, don't worry, I myself have been a bit occupied recently and couldn't give this PR all the attention that it needs. I am hoping to stretch to the latest release (dear imgui 1.82) next week.

I am willing to help and take some responsibilities if it can help this project moving forward. But I am not sure if I have all the requested skills to maintain all of it by myself. I am not that familiar with the entire building pipeline and didn't touch much of the documentation either (else than the addition of docstring to mapped functions). As you suggested, it would be great to form a small team of contributors to organize and help each other.

Concerning this push request, as said two paragraphs earlier, I plan to stretch to the latest imgui release. Probably next week, I am a bit loaded with work this week. After that, I can always touch a bit of code here and there but I have the feeling that this push request is getting out of hand in terms of code size. It is also going in various directions (upgrade + callbacks + internal module + etc) and I think all these things should be focused on one at a time. 

Thus here is what I would suggest: The branch, as it is, is clearly not ready to be pushed in production and a quality pass is needed. Unfortunately having all these changes in one pull request seems not to be very friendly to community modification. Maybe we could merge this PR in a specific 'dev' branch or 'upgrade' branch. By doing this, I think it would be clearer what is the development state of this upgrade and that we are preparing 2.0.0, and contributions could be managed via push request directly. For example, people keep mapping functions that have already been mapped here as in #197 or #218, which is comprehensible since it is difficult, even for myself, to see what has been mapped or not (the list is long ^^).

The mapping of functions is pretty straightforward, I tested the majority of them, especially tricky ones, as I mapped them. And as I understand it, some testing is done by using documentation code. Thus I feel this part is pretty solid, maybe going through all of them at some point to be sure. 

Parts that most concerns me are:

* Change in callbacks mechanism: I have tested it as much as possible and I don't see any obvious memory leaks. But we are never free of a tricky python combination of callables that I am not aware of. This part calls for more systematic testing and code review
* Change in build config: I changed a little bit the build config at some point to separate and map the 'internal' submodule. Reviewing and testing this part would be great
* Various fixes in integration code should be reviewed too.
* And as always: better documentation.

Thus I would suggest a merging of this PR in a dev-specific branch for organizing and attacking these reviews and quality pass separately. If other contributors want to push request stuff for the next major release they would be able to do it in a centralized branch.

What do you think?

learn-more commented:
I would suggest to freeze it, and not keep following every Dear ImGui upgrade,
otherwise this PR will be a moving target.

hinxx commented:
I've been using this PR (v1.81) over the upstream for a while now. Looking at the latest comments it seems lots of things happening, including the upstream maintainer-ship being on the table. I do not have any input on how to tackle this, but would like to see some sort of merging of this effort into either upstream or this repo in order not to lose this huge amount of work.

I agree with @learn-more that if the goal posts with merging this keep moving, it might become endless journey. I feel that v1.80 was a huge jump on ImGui side and now with v1.82 things might have been cleaned up just enough to allow merging this PR.

FWIW, I'm working on having ImPlot python binding (actually have it in alpha as I write this) and it heavily relies on this PR. Not sure how to actually release it: a) separate repo or b) ask you guys if you thing it could be included here as part of pyimgui. Sorry for for being off topic.

Thanks!

KinoxKlark commented:
> I would suggest to freeze it, and not keep following every Dear ImGui upgrade,
> otherwise this PR will be a moving target.

I agree with you. This PR should be kept stable and be finalized. The keep up with ImGui releases should be done in parallel. I think it is important to have a branch that is kept up to date with official releases, from my experience in this upgrade, going from one release to the other is a manageable amount of work but skipping several releases as it was at first (from v1.65 to v1.79) is a huge investment. 

I am not sure how we should proceed from now, maybe we should freeze it in v1.82 and ensure this is stable enough and merge with other PR to release a 2.0.0 version of pyimgui as it was originally suggested by @swistakm. This for now entirely depends on @swistakm's spare time and, as I understand it, it is unfortunately not for soon.

> FWIW, I'm working on having ImPlot python binding (actually have it in alpha as I write this) and it heavily relies on this PR. Not sure how to actually release it: a) separate repo or b) ask you guys if you thing it could be included here as part of pyimgui. Sorry for for being off topic.

Yeah! Sounds awesome. I am not sure what to say concerning its release. In my opinion, 'PyImGui' should be kept as simple as the official 'DearImGui' and allow external extensions. The question sounds similar to the one of the docking branch in the official repo and how to deal with it.

pvallet commented:
Hello! I have great interest in this PR, notably for what @hinxx is working on (I'd love to have implot bindings). Do we have any updates on what's next? Are you all just waiting for @swistakm to give access to the repository to be able to merge that PR upstream, or does the PR still require some work?

KinoxKlark commented:
Hello, glad to hear about your interest in this PR ! Unfortunately, I have been loaded with work for the end of the semester at my university, I had to do a pause to focus on other projects and I won't be able to continue this until next month. As for now, this PR is functioning but requires more testing to be used as a release. Many minor functionalities can be added (e.g. mapping of functions), but pyimgui have never been fully mapped thus what has already be done could be enough for now. As you mentioned @hinxx was working on an external binding of implot, to this respect we discussed how to share the cpp imgui lib to external modules here: https://github.com/KinoxKlark/pyimgui/pull/2 . I think this is temporarily in pause too but definitely a cool project ^^ 

For the next, we are waiting on @swistakm to be able to move forward, although there is no clear plan of what to do. There is also a new release of imgui (v1.83) with new functionalities to map but I don't know if we should do it here or start a new PR.

hinxx commented:
I'm pretty much in the same situation with regards to lack of time in the coming weeks.

I still do have interested in going forward with pyimplot and have something that could be tested and evaluated right now but needs a bit of coordination to make us all comfortable in using it.

I see myself devoting more time to this after summer vacation is over.

TTimo commented:
fwiw we're using the branch in a small internal app, and haven't had any problems with it

KinoxKlark commented:
Hello, I will finally have more time to invest in this project! 

I am not really sure about how we should continue here. In the absence of @swistakm I don't think there is much we can do concerning the official release. For now, I think the best we can do is to make this pr as robust as possible so that people interested in a stable branch with more up-to-date imgui functionalities could use it by manual installation.

I will need little time to go on stuff that I have missed the last few months. I would appreciate some insight about the following matters and how we should deal with them:
- Should we merge here fixes from other pr ? (e.g. #232)
- A new release of dear imgui is here, 1.83, what should we do about that? As it was mentioned earlier, it may be not a good idea to indefinitely chasse dear imgui releases. Maybe we should settle to 1.82 and stabilize pyimgui enough. Should we create another branch (and another pr?) for the implementation of new stuff and upgrade to the latest dear imgui?
- @hinxx and I were discussing potential changes on the distribution of the cpp dearimgui with pyimgui such that it would be possible for an external module to share a unique implementation of dearimgui making it possible to create extensions (like implot). I did not wrap my head around it for a long time, where were we on that matter and how should we consider it here (if we should)?
- I did not have time yet to investigate it deeply but it seems to be an interest for new functionalities like in #233, or questions about the core functioning of pyimgui like in #234 (from way before I came around here ^^). What should we do about that?

I may have forgotten some important stuff, feel free to add to that list.

My hope is that we would be able to maintain from this pr a good enough version to be used by anyone who needs it. And maybe one-day @swistakm would manage to merge it and drop a stable release quickly.

Thank's in advance, looking forward to your advice!

pvallet commented:
To me it feels like in order to keep this project alive and healthy we really need an official place to have it. Incremental changes to a PR, or several pending PRs adding all new functionality is not a good base for welcoming new members of a community or maintaining a stable code base.

Obviously what would be optimal is to have @swistakm back to help maintaining the project or to give admin rights to the ones that are the most involved in maintaining that project already. (I guess @KinoxKlark and @hinxx ? Not sure who I'm missing) This way everything could be stabilized properly, merged to master and then we could fix CI issues and publish to PyPI. It would also allow merging further fixes easily, maintain the documentation and so on

In the meantime, I guess what could be best is doing the same in a fork, like maybe yours @KinoxKlark?

Froyok commented:
I have been working on my own projects with this PR for a while now and I can vouch for it as well. I made UI focused app and other with OpenGL render interactions via SDL and didn't get into any troubles.

Only issue I found so far it is a little keyboard input issue. :)
(The shortcut **CTRL+A** -select all- in text input fields doesn't work. Other shortcut such as **CTRL+Arrows** -word jump-, **Shift+Key** -uppercase- or **SHIFT+Arrows** -character selection- work fine.

swistakm commented:
Hey all! I really apologize for not being responsive in this project. I hoped that once I will finish my book I will have some time, but it turns out I will  probably be eternally busy and unable to properly support such projects. The least I can do is to handverthe project to someone (or group of people). I may be able to support project with other resource but probably won't have time for active development anymore.

I know that we had this conversation before and we didn't decide on anything but now I think it is the only solution.

Probably that's simply not a thing that can be easily discussed asynchronously via GitHub issues. Best if we could meet online (e.g. via Zoom) and discuss this live.

@KinoxKlark, @TTimo, @hinxx: I was thinking mostly about three of you but anyone who want's to actively participate and has some spare time is free to join. If you are interested please drop me an email (address is in my Github). If you give me your time availability ( in UTC) I will setup a call where we can finally meet and discuss how to move forward. Best if we could meet over weekend because next week I leave my country for a couple of days.

learn-more commented:
Might I suggest to transfer the repo to an organization, where multiple people can join in on the fun?

or if you want to keep the repo under your name, make a few people able to add / remove participants on this repo.

swistakm commented:
@learn-more organization of course

KinoxKlark commented:
Thanks all for your answers! 

Nice to hear from you @swistakm. I'll drop you a mail to organize a meeting, it seems more appropriate indeed. I accept to take more responsibilities if it is needed to move forward. I agree with @learn-more, I think we need to make sure multiple people have access and can manage the project in the case where one of us can't be present.

@Froyok, thank you a lot for your feedback. I'll check on that issue as soon as possible ;)

KinoxKlark commented:
Hey everyone,

You may have noticed that the project is now an organization one. Moreover, I now have admin right to the repo (and organization) thanks to @swistakm. I opened a discussion about the future and management of the project here: #237. I hope to gather your opinions there. We will be able to move forward with the project and prepare future releases!

There are still details to manage about the succession (like PyPi access for distribution). Those will be dealt with as we go along.

I will take some vacation in the coming week so I won't be always present for the development. I will still try to come regularly for the discussion. Feel free to comments in #237 and gives any insights/advice. Thanks in advance!
